spec/lib/signon/client_spec.rb

```diff
@@ -79,8 +83,80 @@
     end
   end
 
-  def stub_req(endpoint)
-    stub_request(:post, endpoint)
+  describe "#create_application" do
+    subject(:response) do
+      client.create_application(
+        name: "app",
+        description: "app desc",
+        home_uri: "https://app.example.org",
+        permissions: %w[signin],
+        redirect_uri: "https://app.example.org/gds/auth/callback",
+      )
+    end
+
+    let(:endpoint) { "#{api_url}/applications" }
+
+    context "when signon request is successful" do
+      let(:res) { { "oauth_id" => "a", "oauth_secret" => "b" } }
+
+      it "does not raise an error" do
+        stub_req(endpoint).to_return(
+          status: 200, body: JSON.generate(res),
+        )
+        expect { response }.not_to raise_error
+        expect(response).to eq res
+      end
+    end
+
+    context "when application already exists" do
+      it "raises a custom error" do
+        stub_req(endpoint).to_return(
+          status: 400,
+          body: JSON.generate(error: "Record already exists"),
+        )
+        expect { response }.to raise_error(Signon::Client::ApplicationAlreadyCreated)
+      end
+    end
+
+    context "when signon request fails" do
+      it "raises a custom error" do
+        stub_req(endpoint).to_return(
+          status: 400,
+          body: JSON.generate(error: "Invalid request"),
+        )
+        expect { response }.to raise_error(Signon::Client::ApplicationNotCreated)
+      end
+    end
+  end
+
+  describe "#get_application" do
+    subject(:response) do
+      client.get_application(name: "[Workspace] Publishing API")
+    end
+
+    let(:endpoint) { "#{api_url}/applications?name=%5BWorkspace%5D+Publishing+API" }
+
+    context "when signon request is successful" do
+      let(:res) { { "oauth_id" => "a", "oauth_secret" => "b" } }
+
+      it "does not raise an error" do
+        stub_req(endpoint, method: :get)
+          .to_return(status: 200, body: JSON.generate(res))
+        expect { response }.not_to raise_error
+        expect(response).to eq res
+      end
+    end
+
+    context "when signon request fails" do
```

Comment:
"when an application isn't found"? (the description is currently the same as the 400 case above)